### DISCUSSION TITLE
Wrong Bitcoin addresses generating for legacy, SegWit and Native SegWit (BIP44|49|84)

### DISCUSSION THREAD

**Discussion by EvgenV** [trustwallet/wallet-core#2778] [Q&A]
I tried to test possibility to generate additional addresses for different versions of Bitcoin's addresses.
I decided to use `final public func getKey(coin: WalletCore.CoinType, derivationPath: String) -> WalletCore.PrivateKey`
So idea is to specify custom index, e.g. `"m/44\'/0\'/0\'/0/2"` or custom account `"m/44\'/0\'/2\'/0/0"`

But I was stuck at the generating default address for legacy format. 
I use `Swift`

```
let mnemonicPhrase = "foster shallow problem poverty similar theme rule ignore camp van chuckle odor" // Metamask
guard let wallet = HDWallet(mnemonic: mnemonicPhrase, passphrase: "") else {
    throw HDWalletError.impossibleToGenerateWallet
}
let bitcoinDerivationPathLegacy = "m/44\'/0\'/0\'/0/0"
let bitcoinKeyLegacy = wallet.getKey(coin: .bitcoin, derivationPath: bitcoinDerivationPathLegacy)
let bitcoinAddressLegacy = CoinType.bitcoin.deriveAddress(privateKey: bitcoinKeyLegacy)
print("\(bitcoinDerivationPathLegacy): \(bitcoinAddressLegacy)")
// m/44'/0'/0'/0/0: bc1qy3laaxl4f9022vn49s0a3ptnz25pj967l96g5e

let bitcoinDerivationPathSegwit = "m/49\'/0\'/0\'/0/0"
let bitcoinKeySegwit = wallet.getKey(coin: .bitcoin, derivationPath: bitcoinDerivationPathSegwit)
let bitcoinAddressSegwit = CoinType.bitcoin.deriveAddress(privateKey: bitcoinKeySegwit)
print("\(bitcoinDerivationPathSegwit): \(bitcoinAddressSegwit)")
// m/49'/0'/0'/0/0: bc1qfv0xyycjr59rfhdy4k48ch039p442hh3j33vd4

let bitcoinDerivationPathNativeSegwit = "m/84\'/0\'/0\'/0/0"
let bitcoinKeyNativeSegwit = wallet.getKey(coin: .bitcoin, derivationPath: bitcoinDerivationPathNativeSegwit)
let bitcoinAddressNativeSegwit = CoinType.bitcoin.deriveAddress(privateKey: bitcoinKeyNativeSegwit)
print("\(bitcoinDerivationPathNativeSegwit): \(bitcoinAddressNativeSegwit)")
// m/84'/0'/0'/0/0: bc1qaky53wlhtfvm6zup8hhfr3vcg2uvg23mrs9p7s
```

I expected that address generated with `derivationPath m/44'/0'/0'/0/0` starts from `1`
address generated with `derivationPath m/49'/0'/0'/0/0` starts from `3`

What I am doing wrong?

**Comment by Milerius** [comment]
Converting to discussion as it's a question

**Reply by optout21** [reply]
Address derivation has these steps:
1. Generating the private key from the mnemonic (based on a derivation path)
2. public key from private key
3. creating a string encoding from the public key

Both 1 and 3 differ in different address schemes (2 is constant).
With the different derivation paths you changed only step 1.
Wallet core supports so called Legacy and Segwit addresses for Bitcoin, as well as Testnet (more recently).

Please consult these items/tests:

https://github.com/trustwallet/wallet-core/blob/master/registry.json#L9
https://github.com/trustwallet/wallet-core/blob/master/tests/chains/Bitcoin/SegwitAddressTests.cpp#L239
https://github.com/trustwallet/wallet-core/blob/master/tests/interface/TWAnyAddressTests.cpp#L185
https://github.com/trustwallet/wallet-core/blob/master/swift/Tests/AnyAddressTests.swift#L43
